libs/game/prompt.ts

```diff
@@ -0,0 +1,541 @@
+
+    // Dimensions of the bottom bar
+    //% whenUsed=true
+    const BOTTOM_BAR_ALPHABET_MARGIN = 4;
+    //% whenUsed=true
+    const BOTTOM_BAR_HEIGHT = PADDING + BOTTOM_BAR_ALPHABET_MARGIN + CELL_HEIGHT;
+    //% whenUsed=true
+    const BOTTOM_BAR_TOP = screen.height - BOTTOM_BAR_HEIGHT;
+    //% whenUsed=true
+    const BOTTOM_BAR_BUTTON_WIDTH = PADDING * 2 + font.charWidth * 3;
+    //% whenUsed=true
+    const BOTTOM_BAR_TEXT_Y = (BOTTOM_BAR_HEIGHT - font.charHeight) / 2;
+    //% whenUsed=true
+    const BOTTOM_BAR_SHIFT_X = (BOTTOM_BAR_BUTTON_WIDTH - font.charWidth * 3) / 2;
+    //% whenUsed=true
+    const BOTTOM_BAR_CONFIRM_X = (BOTTOM_BAR_BUTTON_WIDTH - font.charWidth * 2) / 2;
+    //% whenUsed=true
+    const CONFIRM_BUTTON_LEFT = screen.width - BOTTOM_BAR_BUTTON_WIDTH;
+
+    // Dimensions of the alphabet area
+    //% whenUsed=true
+    const ALPHABET_HEIGHT = NUM_ROWS * CELL_HEIGHT;
+    //% whenUsed=true
+    const ALPHABET_TOP = CONTENT_TOP + CONTENT_HEIGHT - ALPHABET_HEIGHT - BOTTOM_BAR_HEIGHT;
+    //% whenUsed=true
+    const ALPHABET_INPUT_MARGIN = 10;
+
+    // Dimensions of area where text is input
+    //% whenUsed=true
+    const INPUT_HEIGHT = INPUT_ROWS * CELL_HEIGHT;
+    //% whenUsed=true
+    const INPUT_TOP = ALPHABET_TOP - INPUT_HEIGHT - ALPHABET_INPUT_MARGIN;
+
+    // Dimensions of prompt message area
+    //% whenUsed=true
+    const PROMPT_HEIGHT = INPUT_TOP - CONTENT_TOP;
+
+    //% whenUsed=true
+    const lowerShiftText = "ABC";
+    //% whenUsed=true
+    const upperShiftText = "abc";
+    //% whenUsed=true
+    const digitsUpper = [" ", ",", ".", "?", "!", ":", ";", "\"", "(", ")"];
+    //% whenUsed=true
+    const confirmText = "OK";
+
+
+    export class Prompt {
+        theme: PromptTheme;
+
+        message: string;
+        answerLength: number;
+
+        cursor: Sprite;
+        shiftButton: Sprite;
+        confirmButton: Sprite;
+
+        letters: Sprite[];
+        inputs: Sprite[];
+
+        confirmPressed: boolean;
+        cursorRow: number;
+        cursorColumn: number;
+        upper: boolean;
+        inputIndex: number;
+        result: string;
+        blink: boolean;
+        frameCount: number;
+
+        constructor(theme?: PromptTheme) {
+            if (theme) {
+                this.theme = theme;
+            }
+            else {
+                this.theme = {
+                    colorPrompt: 1,
+                    colorInput: 3,
+                    colorInputHighlighted: 5,
+                    colorInputText: 1,
+                    colorAlphabet: 1,
+                    colorCursor: 7,
+                    colorBackground: 15,
+                    colorBottomBackground: 3,
+                    colorBottomText: 1,
+                };
+            }
+            this.cursorRow = 0;
+            this.cursorColumn = 0;
+            this.upper = false;
+            this.inputIndex = 0;
+        }
+
+        _draw() {
```

Comment:
We should really have a consistent naming convention. Should this be a private function?

Reply:
I think so, it's the internal drawing function. Really the only public API on this class right now is `show()`